[crystal] add init command

waterlink opened the pull request:
closes #505 

```
➜  tmp  ../crystal/.build/crystal init --help
USAGE: crystal init TYPE NAME [DIR]

TYPE is one of:
    lib                      creates library skeleton
    app                      creates application skeleton

NAME - name of project to be generated,
       eg: example
DIR  - directory where project will be generated,
       default: NAME, eg: ./custom/path/example

    --help                           Shows this message
```

Generates following project structure:

```
➜  crystal  crystal init lib example
      create  example/.gitignore
      create  example/LICENSE
      create  example/README.md
      create  example/.travis.yml
      create  example/Projectfile
      create  example/src/example.cr
      create  example/src/example/version.cr
      create  example/spec/spec_helper.cr
      create  example/spec/example_spec.cr
Initialized empty Git repository in /Users/your_user_name/crystal/example/.git/
➜  crystal  cd example
➜  example  tree . -a
.
├── .git
│   ├── HEAD
# ...
├── .gitignore
├── .travis.yml
├── LICENSE
├── Projectfile
├── README.md
├── spec
│   ├── example_spec.cr
│   └── spec_helper.cr
└── src
    ├── example
    │   └── version.cr
    └── example.cr
```

- license file specifies MIT, is it OK ?
- I have added this `src/example/version.cr` which is not used currently by anything, but I expect, that it will come in handy at some point.
- anything you think needs to be added to git ignore ?
- maybe we want some output of this command (like which files and directories got generated) ? - now it is silent.

WDYT?  /cc @asterite

jhass commented on src/crystal/init_project/template/gitignore.ecr:
```diff
@@ -0,0 +1,4 @@
+/.deps/
+/.deps.lock
```
I think we want to keep the .lock in the repo, at least I understood it as having the same functionality as bundlers `Gemfile.lock`.

waterlink replied:
If you generate gem with `bundle gem`, then `Gemfile.lock` is in `.gitignore`.

This is because for libraries it is OK to break on newest versions (and you want to know about that, for example on travis build). Because they will be locked in applications that use them as a dependency.

On the other hand, bundler and rubygems do have gemspec (which actually is used for locking down versions, at least constraining them) - which is committed to source control.

Applications on the other hand will definitely want to lock down the versions and commit `.deps.lock`.

waterlink replied:
So the question is, what we want here as a default? Library or Application behavior of git ignore?

asterite replied:
Maybe we can have `crystal init lib example` and `crystal init app example`. If you write `crystal init` it will tell you to choose `app` or `lib`.

alex-fedorov commented:
@asterite @jhass I have made as suggested:

- Added 2 project skeleton types: `app` and `lib`
- Used `OptionParser` for `InitCli.run` to show nice help message.

plus, I have removed `src/example/.keep`, since it is not needed actually.

asterite commented:
@alex-fedorov Awesome!!

I just looked at the code and it's really good. I have these comments:

1. Right now all "tools" (hierarchy, type, browser, doc) are placed in `src/compiler/crystal/tools`, could you put `init` there? Just to keep things ordered.
2. The other comments I'll do inline in the code :-)

Thanks!!

asterite commented on src/crystal/init_cli.cr:
```diff
@@ -0,0 +1,141 @@
+module Crystal
+  module InitCli
+    def self.run(args)
+      options = {} of Symbol => String
```
Instead of using a Hash, maybe you could do `config = Config.new` and let config have properties instead of just getters. Then the option parser would set those properties and pass it directly to `InitProject`. That way we avoid an intermediate Hash.

(I don't think this will affect performance at all, the Hash is small, but using `Config` directly will make things a bit more type safe... and in any case all of its properties are nilable)

asterite commented on src/crystal/init_cli.cr:
```diff
@@ -0,0 +1,141 @@
+        end
+      end
+
+      options[:author] = `git config --get user.name`.strip
```
Do you know what happens if `git` is not installed? Maybe we should check that case too.

waterlink replied:
Yeah, here now I'm checking if it is available on PATH with `which`, and if it is not, then I provide a placeholder `[your-name-here]`

asterite commented on src/crystal/init_cli.cr:
```diff
@@ -0,0 +1,141 @@
+
+    def self.fetch_skeleton_type(opts, args)
+      skeleton_type = fetch_required_parameter(opts, args, "TYPE")
+      unless ["lib", "app"].includes?(skeleton_type)
```
Here you can do:

```crystal
unless {"lib", "app"}.includes?(skeleton_type)
```

This avoids an array allocation :-)

(yes, I know, performance obsession :-P)

asterite commented:
And to your other questions:

* MIT is OK, later they can change it
* `version.cr` is OK, maybe later it'll be useful
* Output: yes, maybe it would be nice for the command to say what it has created, similar to `bundle gem`. Or at least something like "Created app/lib at #{full_path_to_app}", maybe that'll be enough.

waterlink commented:
Ah, actually made exactly the same output as a bundler (except terminal coloring for "create" word).

And addressed all other issues.

And added automatic `git init` if `git` is available (bundler does it too)

waterlink commented:
You can see example output in example in description to PR

waterlink commented:
/cc @asterite @jhass

waterlink commented:
Randomly stumbled upon `Object#colorize`, so now `create` word does have light green color :)

kostya commented on src/compiler/crystal/tools/init.cr:
```diff
@@ -0,0 +1,181 @@
+    template SrcVersionView, "version.cr.ecr", "src/#{config.name}/version.cr"
+
+    template SpecHelperView, "spec_helper.cr.ecr", "spec/spec_helper.cr"
+    template SpecExampleView, "example_spec.cr.ecr", "spec/#{config.name}_spec.cr"
```
list of views described in two places, is there any way, to auto generate method `views`

waterlink replied:
Then I would probably need to have instances of `ViewFactory` in the `views` array. Currently I have an array of subclasses (not instances). I don't even know how to create empty array of this thing.

jhass replied:
I guess we could have `VIEWS = [] of Class` and the macro do `VIEWS << {{name}}` or the `View` class have a `macro inherited` which does that. I'm not sure if it's really worth it though.

asterite replied:
You could use this:

```crystal
abstract class Foo
end

class Bar < Foo
end

class Baz < Foo
end

macro def foo_subclasses : Array(Foo.class)
  {{ Foo.subclasses }}
end

p foo_subclasses #=> [Bar, Baz]
```

It still can't be written generically, as in `Foo.subclasses` because of some limitations/bugs.

asterite replied:
(but I don't mind the small repetition)

waterlink replied:
Wow.

In the meantime I came up with this: https://github.com/waterlink/crystal/commit/6675b3f4dd4719c8ad38a354433d105de823dd56

Old good factory class approach.

waterlink replied:
Last time when I tried to do `SomeClass.class` and use it as a type of `Array(...)` I got scary compiler backtraces :D

waterlink replied:
Actually `@@views = Array of View.class` works just fine, so no factory required: https://github.com/waterlink/crystal/commit/5c26f35e477b4c8cf2df790e53c10a034c43a60a  (diff is rendered strange, since I moved class)

waterlink replied:
So WDYT?:

- leave it with a bit of duplication?
- go with `macro view_subclasses` ?
- go with `Array of View.class` and `#register_view` ?

asterite replied:
I think the way you left it is good (the one with register_view) :-)

jhass replied:
I actually wouldn't mind the duplication too much, but if we DRY it up, the `view_subclasses` approach seems most reasonable.

asterite replied:
Yes, I think any approach is fine (even the one with duplication). This code won't change that much later, and if it changes we'll have to test it and see if the newly added content is there (and there are also specs).

waterlink commented:
@asterite On the side note about general subclasses thingy:

```crystal
macro enable_subclasses(target)
  macro \{{target.id}}_subclasses : Array(\{{target.id}}.class)
    \{{ {{target.id}}.subclasses }}
  end
end
```

even might work (didn't tried), but I know that macro inside of macro works just fine (tried).

waterlink commented:
Ok, removed duplication with `View.register({{name.id}})` approach

waterlink commented:
It failed again with:

```
cat: stdin: Resource temporarily unavailable
make: *** [spec] Broken pipe: 13
```

Is it something that occurred from time to time on mac os x travis build?

waterlink commented:
Or is it something my change introduced?

asterite commented:
@waterlink It's something broken in travis, I think. It happens from time to time, I didn't investigate why. I'll merge your changes very soon :-)

asterite commented:
@waterlink I just tried it, works like a charm!! Thank you!! :heart: :yellow_heart: :green_heart: :blue_heart:
